pkg/kubelet/network/cni/cni.go

```diff
@@ -0,0 +1,232 @@
+		conf, err := libcni.ConfFromFile(confFile)
+		if err != nil {
+			glog.Warningf("Error loading CNI config file %s: %v", confFile, err)
+			continue
+		}
+		// Search for vendor-specific plugins as well as default plugins in the CNI codebase.
+		vendorCNIDir := fmt.Sprintf(VendorCNIDirTemplate, conf.Network.Name)
+		cninet := &libcni.CNIConfig{
+			Path: []string{DefaultCNIDir, vendorCNIDir},
+		}
+		networkMap[conf.Network.Name] = &cniNetwork{name: conf.Network.Name, NetworkConfig: conf, CNIConfig: cninet}
+	}
+	return networkMap
+}
+
+func (plugin *cniNetworkPlugin) selectNetwork(pod *api.Pod) (*cniNetwork, error) {
+	// loading networks all over again maybe inefficient, but required if networks can be created on the fly
+	plugin.cniNetworkMap = getAllCNINetworks()
+
+	if len(plugin.cniNetworkMap) == 0 {
+		return nil, fmt.Errorf("No available CNI network available in %s", DefaultNetDir)
+	}
+
+	// check if the namespace of the pod is a network name itself
+	network, ok := plugin.cniNetworkMap[pod.Namespace]
+	if ok {
+		return network, nil
+	}
+
+	// label called "network"?
+	network, ok = plugin.cniNetworkMap[pod.Labels["network"]]
+	if ok {
+		return network, nil
+	}
+
+	// annotation called "network"?
+	network, ok = plugin.cniNetworkMap[pod.Annotations["network"]]
+	if ok {
+		return network, nil
+	}
+
+	// else return the first one
+	for _, network := range plugin.cniNetworkMap {
+		return network, nil
+	}
+
+	return nil, fmt.Errorf("No available CNI network available in %s", DefaultNetDir)
+}
+
+func (plugin *cniNetworkPlugin) Init(host network.Host) error {
+	plugin.host = host
+	return nil
+}
+
+func (plugin *cniNetworkPlugin) Name() string {
+	return CNIPluginName
+}
+
+func (plugin *cniNetworkPlugin) SetUpPod(namespace string, name string, id kubeletTypes.DockerID) error {
+	netns, err := plugin.host.GetRuntime().GetNetNs(string(id))
+	if err != nil {
+		return err
+	}
+
+	pod, ok := plugin.host.GetPodByName(namespace, name)
+	if !ok {
+		return fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	// TODO: pick one network? all networks? which one writes to Status.PodIP?
+	// Perhaps pick the network by some annotation/label on pod, or the namespace.
+	network, err := plugin.selectNetwork(pod)
+	if err != nil {
+		return err
+	}
+	res, err := network.addNetwork(name, namespace, string(id), netns)
+
+	var ip string
+	if res.IP4 != nil {
+		ip = res.IP4.IP.String()
+	} else {
+		ip = res.IP6.IP.String()
+	}
+	// TODO-PAT: check that PodIP can be an IPv6.
+	pod.Status.PodIP = ip
+	return err
+}
+
+func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id kubeletTypes.DockerID) error {
+	netns, err := plugin.host.GetRuntime().GetNetNs(string(id))
+	if err != nil {
+		return err
+	}
+	pod, ok := plugin.host.GetPodByName(namespace, name)
+	if !ok {
+		return fmt.Errorf("pod %q namespace %q: unable to find pod", name, namespace)
+	}
+
+	network, err := plugin.selectNetwork(pod)
```

Comment:
the downside of re-evaluating the list of networks is that it can change.  Which is worse (for now): saying that the list of networks can't change without a restart of kubelet or "forgetting" to detach from some?